bolt/kv.go

```diff
@@ -233,19 +254,46 @@ func (c *Cursor) Last() ([]byte, []byte) {
 }
 
 // Next retrieves the next key in the bucket.
-func (c *Cursor) Next() ([]byte, []byte) {
-	k, v := c.cursor.Next()
+func (c *Cursor) Next() (k []byte, v []byte) {
+	// get and unset previously seeked values if they exist
+	k, v, c.key, c.value = c.key, c.value, nil, nil
+	if len(k) > 0 && len(v) > 0 {
+		return
+	}
+
+	next := c.cursor.Next
+	if c.config.Direction == kv.CursorDescending {
+		next = c.cursor.Prev
+	}
+
+	k, v = next()
 	if len(k) == 0 && len(v) == 0 {
 		return nil, nil
 	}
```

Comment:
I saw this for the original cursor code and wonder why it doesn't just return the key / value, as the caller is required to check `len(key) == 0` to test if it is at the end again.

Not a required change, but perhaps we can clean this up if we remove the old `Cursor` API.